fiftyone/core/map/threading.py

```diff
@@ -0,0 +1,240 @@
+    Callable,
+    Iterator,
+    List,
+    Literal,
+    Optional,
+    Tuple,
+    TypeVar,
+    Union,
+)
+
+import bson
+import numpy as np
+from tqdm import tqdm
+
+import fiftyone.core.utils as fou
+import fiftyone.core.view as fov
+import fiftyone.core.map.map as fomm
+import fiftyone.core.map.sequential as foms
+
+
+T = TypeVar("T")
+R = TypeVar("R")
+
+
+class ThreadingMapBackend(foms.SequentialMapBackend):
+    """Executes map_samples with threading using iter_samples."""
+
+    def __init__(self, max_shard_size: Union[int, None] = None):
+        self.__max_shard_size = (
+            max_shard_size
+            or fou.recommend_batch_size_for_value(
+                bson.ObjectId(), max_size=100000
+            )
+        )
+
+    def split_sample_collection(
+        self,
+        root_sample_collection: fomm.SampleCollection[T],
+        num_workers: int,
+        shard_method: Literal["id", "slice"],
+        shard_size: Union[int, None] = None,
+    ) -> list[tuple[fomm.SampleCollection[T], int]]:
+        """Split a sample collection into multiple sample collections"""
+
+        if shard_method == "slice":
+            n = len(root_sample_collection)
+        else:
+            ids = root_sample_collection.values("id")
+            n = len(ids)
+
+        # Must cap size of select(ids) stages
+        if shard_method == "id":
+            if shard_size is not None:
+                shard_size = min(shard_size, self.__max_shard_size)
+            elif n > num_workers * self.__max_shard_size:
+                shard_size = self.__max_shard_size
+
+        if shard_size is not None:
+            # Fixed size shards
+            edges = list(range(0, n + 1, shard_size))
+            if edges[-1] < n:
+                edges.append(n)
+        else:
+            # Split collection into exactly `num_workers` shards
+            edges = [int(round(b)) for b in np.linspace(0, n, num_workers + 1)]
+
+        split = []
+        for start, stop in zip(edges[:-1], edges[1:]):
+            total = stop - start
+            if shard_method == "slice":
+                view = root_sample_collection[start:stop]
+                split.append((view, total))
+            else:
+                view = fov.make_optimized_select_view(
+                    root_sample_collection, ids[start:stop]
+                )
+                split.append((view, total))
+
+        return split
+
+    @staticmethod
+    def sample_map_worker(
+        sample_collection: fomm.SampleCollection[T],
+        sample_count: int,
+        batch_num: int,
+        batch_count: int,
+        q: queue.Queue,
+        event: threading.Event,
+        /,
+        map_fcn: Callable[[T], R],
+        save: bool,
+        progress: bool,
+    ):
+        """Map samples"""
+
+        samples = sample_collection.iter_samples(autosave=save)
+
+        if progress:
+            desc = f"Batch {batch_num:0{len(str(batch_count))}}/{batch_count}"
+            samples = tqdm(
```

Comment:
I see you want to use TQDM instead of our own progress bar - is there any specific reason?